### PR TITLE
Core: C++ core

### PR DESCRIPTION
**Pull request by iopapamanoglou**
# Core: C++ Core

# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

# Checklist

Please read and execute the following:

- [ ] My code follows the [coding guidelines](/faebryk/faebryk/blob/main/docs/CODING_GUIDELINES.md) of this project
- [ ] My PR title is following the [contribution guidelines](/faebryk/faebryk/blob/main/docs/CONTRIBUTING.md)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (if applicable)
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I ran [Black](../docs/CONTRIBUTING.md#creating-a-pull-request) to format my code

#### Code of Conduct

By submitting this issue, you agree to follow our [Code of Conduct](/faebryk/faebryk/blob/main/docs/CODE_OF_CONDUCT.md):

- [ ] I agree to follow this project's [Code of Conduct](/faebryk/faebryk/blob/main/docs/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: src/faebryk/core/graph.py [hunk 1]
```diff
@@ -2,142 +2,47 @@
...
-        return self.subgraph(lambda n: isinstance(n, types))
 
-    # TODO remove boilerplate for lazy, if it becomes a bit more usable
-    def __init_subclass__(cls) -> None:
-        super().__init_subclass__()
-        if LAZY:
-            lazy_construct(cls)
+# TODO move these to C++
+# a bit weird typecheck
+class GraphFunctions:
```
Comment (mawildoer) [inline src/faebryk/core/graph.py:18]: This pattern of a seperate not-quite-subclass to wrap functions is a little odd. Is this only as a placeholder to soon refactor, or do you plan to keep this?

File: src/faebryk/core/graphinterface.py [hunk 2]
```diff
@@ -1,220 +1,14 @@
...
-    find,
-    not_none,
-    try_avoid_endless_recursion,
-)
-
-if TYPE_CHECKING:
-    from faebryk.core.node import Node
-    from faebryk.core.trait import Trait
 
 logger = logging.getLogger(__name__)
```
Comment (mawildoer) [inline src/faebryk/core/graphinterface.py:5]: This logger also seems unreferenced now

File: src/faebryk/core/cpp/CMakeLists.txt [hunk 3]
```diff
@@ -38,6 +38,8 @@ message(STATUS "Python_EXECUTABLE: ${Python_EXECUTABLE}")
 message(STATUS "Python_INCLUDE_DIR: ${Python_INCLUDE_DIR}")
 message(STATUS "Python_LIBRARY: ${Python_LIBRARY}")
 
+#set(CMAKE_BUILD_TYPE "Debug")
```
Comment (mawildoer) [inline src/faebryk/core/cpp/CMakeLists.txt:41]: nit; Make this an arg on input, environment var etc... Is this meant to remain a hanging comment?

**Review by mawildoer — COMMENTED** [review]
Can we pull any dependencies (`networkx`) as part of this PR?

**Review by mawildoer — APPROVED** [review]

**Review by mawildoer — CHANGES_REQUESTED** [review]
Ran the tests and three tests hang.

Here's an output of pytest:

```
time pytest
=========================================================================================================================================================================================================== test session starts ============================================================================================================================================================================================================
platform darwin -- Python 3.12.7, pytest-8.3.3, pluggy-1.5.0
rootdir: /Users/mattwildoer/Projects/faebryk
configfile: pyproject.toml
testpaths: test
plugins: dash-2.18.1, xdist-3.6.1
8 workers [532 items]   
........................................................................................................................................................................................................................................................................................................................s...s..s..........s....s........s.ssssss..................................ss...ss.....ss.............s...... [ 78%]
.....s..............s..............................................................s...............s...........^C
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
/opt/homebrew/Cellar/python@3.12/3.12.7_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/threading.py:359: KeyboardInterrupt
(to show a full traceback on KeyboardInterrupt use --full-trace)
=============================================================================================================================================================================================== 508 passed, 23 skipped in 574.77s (0:09:34) ================================================================================================================================================================================================
pytest  755.45s user 19.38s system 134% cpu 9:35.34 total
```

Seems like it also took almost 10m in CI: https://github.com/atopile/faebryk/actions/runs/11708704697/job/32611086600?pr=113

Running again with the just `test_init_basic` till completion was also almost 10 minutes.

Screenshot 2024-11-06 at 15 04 54

File: src/faebryk/core/graph.py (hunk 1 above)
Comment (iopapamanoglou) [reply src/faebryk/core/graph.py:18]: Just placeholder to make it quick and easy to refactor old code

File: src/faebryk/core/cpp/CMakeLists.txt (hunk 3 above)
Comment (iopapamanoglou) [reply src/faebryk/core/cpp/CMakeLists.txt:41]: it already is a CMake arg

File: src/faebryk/core/cpp/CMakeLists.txt [hunk 4]
```diff
@@ -56,7 +58,16 @@ if(${EDITABLE} AND CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
 endif()
 # source files ---------------------------------------------------------
 include_directories(${CMAKE_SOURCE_DIR}/include)
-set(SOURCE_FILES src/main.cpp)
+set(SOURCE_FILES
```
[the remark this answers is not in the document]
Comment (iopapamanoglou) [reply src/faebryk/core/cpp/CMakeLists.txt:61]: Good catch! I like explicit sources better for now. Not sure whether we are going for conditional source files at some point. But I'm open for suggestions.

**Comment by iopapamanoglou** [comment]
@mawildoer 
> Ran the tests and three tests hang.

Just takes a long time.
Connect is very slow. But no reason to waste time on this since the lazy mifs fix this anyway.

> Can we pull any dependencies (`networkx`) as part of this PR?

Good idea!

**Review by mawildoer — APPROVED** [review]

**Pull request merged**